File: src/steamdl_app/lib/api_client.dart
// 引擎 /api 契约的 HTTP 客户端(127.0.0.1 IPC)。
// 使用 dart:io HttpClient,无第三方依赖。
import 'dart:convert';
import 'dart:io';

import 'models.dart';

class ApiException implements Exception {
  final int status;
  final String message;

  ApiException(this.status, this.message);

  @override
  String toString() => message;
}

class ApiClient {
  final String host;
  final int port;
  final HttpClient _client = HttpClient()
    ..connectionTimeout = const Duration(seconds: 5);

  ApiClient({this.host = '127.0.0.1', this.port = 8630});

  Uri _uri(String path, [Map<String, String>? query]) => Uri(
      scheme: 'http',
      host: host,
      port: port,
      path: path,
      queryParameters: query);

  Future<Map<String, dynamic>> _request(
    String method,
    String path, {
    Map<String, dynamic>? body,
    Map<String, String>? query,
    Duration timeout = const Duration(seconds: 20),
  }) async {
    final req =
        await _client.openUrl(method, _uri(path, query)).timeout(timeout);
    if (body != null) {
      req.headers.contentType = ContentType.json;
      req.add(utf8.encode(jsonEncode(body)));
    }
    final res = await req.close().timeout(timeout);
    final text = await utf8.decodeStream(res).timeout(timeout);
    Map<String, dynamic> data = const {};
    if (text.isNotEmpty) {
      try {
        final parsed = jsonDecode(text);
        if (parsed is Map<String, dynamic>) data = parsed;
      } catch (_) {
        // 非 JSON 响应按空对象处理
      }
    }
    if (res.statusCode < 200 || res.statusCode >= 300) {
      throw ApiException(res.statusCode,
          (data['error'] as String?) ?? 'HTTP ${res.statusCode}');
    }
    return data;
  }

  Future<Map<String, dynamic>> _get(
    String path, {
    Map<String, String>? query,
    Duration timeout = const Duration(seconds: 20),
  }) =>
      _request('GET', path, query: query, timeout: timeout);

  Future<Map<String, dynamic>> _post(String path,
          [Map<String, dynamic>? body]) =>
      _request('POST', path, body: body ?? const {});

  // ---- 健康检查 ----

  Future<bool> ping({Duration timeout = const Duration(seconds: 1)}) async {
    try {
      await _get('/api/config', timeout: timeout);
      return true;
    } catch (_) {
      return false;
    }
  }

  Future<Map<String, dynamic>> config() => _get('/api/config');

  Future<Map<String, dynamic>> diagnosticsLog() => _get('/api/diagnostics/log');

  // ---- 账号 ----

  Future<(List<String>, List<AccountDetail>, LoginState, String)>
      accounts() async {
    final data = await _get('/api/accounts');
    final names = ((data['accounts'] as List?) ?? const [])
        .map((x) => x.toString())
        .toList();
    final details = ((data['account_details'] as List?) ?? const [])
        .whereType<Map<String, dynamic>>()
        .map(AccountDetail.fromJson)
        .toList();
    final login = data['login'] is Map<String, dynamic>
        ? LoginState.fromJson(data['login'] as Map<String, dynamic>)
        : LoginState.idle;
    return (names, details, login, (data['selected_account'] ?? '').toString());
  }

  Future<LoginState> login(
      String username, String password, bool rememberPassword) async {
    final data = await _post('/api/accounts/login', {
      'username': username,
      'password': password,
      'remember_password': rememberPassword,
    });
    return LoginState.fromJson(data);
  }

  Future<LoginState> loginInput(String answer) async {
    final data = await _post('/api/accounts/login/input', {'answer': answer});
    return LoginState.fromJson(data);
  }

  Future<LoginState> relogin(String username) async {
    final data = await _post('/api/accounts/relogin', {'username': username});
    return LoginState.fromJson(data);
  }

  Future<void> logout(String username) =>
      _post('/api/accounts/logout', {'username': username});

  Future<void> selectAccount(String username) =>
      _post('/api/accounts/select', {'username': username});

  // ---- 解析与游戏信息 ----

  Future<ParsedTarget> parse(String url) async {
    final data = await _post('/api/parse', {'url': url});
    return ParsedTarget.fromJson(data);
  }

  Future<AppInfo?> appInfo(String appId) async {
    try {
      final data = await _get('/api/appinfo/$appId');
      return AppInfo.fromJson(data);
    } on ApiException {
      return null; // 商店无信息不影响下载
    }
  }

  // ---- 游戏库 ----

  Future<LibraryStatus> libraryStatus(String username) async {
    final data =
        await _get('/api/library/status', query: {'username': username});
    return LibraryStatus.fromJson(data);
  }

  Future<LibraryStatus> librarySync(String username,
      {required bool full}) async {
    final data = await _post('/api/library/sync', {
      'username': username,
      'mode': full ? 'full' : 'incremental',
      'force_full_sync': full,
    });
    return LibraryStatus.fromJson(data);
  }

  // ---- 任务 ----

  Future<List<Job>> jobs() async {
    final data = await _get('/api/jobs');
    return ((data['jobs'] as List?) ?? const [])
        .whereType<Map<String, dynamic>>()
        .map(Job.fromJson)
        .toList();
  }

  Future<Job?> job(String jobId) async {
    try {
      final data = await _get('/api/jobs/$jobId');
      return Job.fromJson(data);
    } on ApiException {
      return null;
    }
  }

  Future<Job> createJob({
    required String kind,
    required String id,
    required String username,
    required String os,
    String depot = '',
    String outputDir = '',
    String installDir = '',
    String name = '',
  }) async {
    final data = await _post('/api/jobs', {
      'kind': kind,
      'id': id,
      'username': username,
      'anonymous': false,
      'os': os,
      'depot': depot,
      'output_dir': outputDir,
      'install_dir': installDir,
      'installdir': installDir,
      'name': name,
    });
    return Job.fromJson((data['job'] as Map<String, dynamic>?) ?? data);
  }

  Future<void> jobAction(String jobId, String action) =>
      _post('/api/jobs/$jobId/$action');

  Future<void> jobInput(String jobId, String answer) =>
      _post('/api/jobs/$jobId/input', {'answer': answer});

  Future<void> deleteJob(String jobId, {required bool deleteFiles}) =>
      _request('DELETE', '/api/jobs/$jobId',
          body: {'delete_files': deleteFiles});

  Future<void> deleteAllJobs({required bool deleteFiles}) =>
      _request('DELETE', '/api/jobs', body: {'delete_files': deleteFiles});

  // ---- 设置 ----

  Future<AppSettings> settings() async =>
      AppSettings.fromJson(await _get('/api/settings'));

  Future<AppSettings> saveSettings(AppSettings settings) async =>
      AppSettings.fromJson(await _post('/api/settings', settings.toJson()));

  void close() {
    _client.close(force: true);
  }
}

File: src/SteamDl.Maui/Services/ApiClient.cs
using System.Net;
using System.Text;
using System.Text.Json;
using SteamDl.Maui.Models;
using SteamAppInfo = SteamDl.Maui.Models.AppInfo;

namespace SteamDl.Maui.Services;

public sealed class ApiException(int status, string message) : Exception(message)
{
    public int Status { get; } = status;
}

public sealed class ApiClient : IDisposable
{
    static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web)
    {
        PropertyNameCaseInsensitive = true,
    };

    readonly HttpClient _client;

    public ApiClient(string host = "127.0.0.1", int port = 8630)
    {
        Host = host;
        Port = port;
        _client = new HttpClient
        {
            BaseAddress = new Uri($"http://{host}:{port}"),
            Timeout = TimeSpan.FromSeconds(20),
        };
    }

    public string Host { get; }
    public int Port { get; }

    public async Task<bool> PingAsync(TimeSpan? timeout = null)
    {
        using var cts = new CancellationTokenSource(timeout ?? TimeSpan.FromSeconds(1));
        try
        {
            await GetAsync<JsonElement>("/api/config", cts.Token);
            return true;
        }
        catch
        {
            return false;
        }
    }

    public Task<JsonElement> ConfigAsync(CancellationToken cancellationToken = default) => GetAsync<JsonElement>("/api/config", cancellationToken);
    public Task<DiagnosticsResponse> DiagnosticsLogAsync(CancellationToken cancellationToken = default) => GetAsync<DiagnosticsResponse>("/api/diagnostics/log", cancellationToken);
    public async Task<AccountsResponse> AccountsAsync(CancellationToken cancellationToken = default) => await GetAsync<AccountsResponse>("/api/accounts", cancellationToken) ?? new([], [], LoginState.Idle, "");
    public Task<LoginState> LoginAsync(string username, string password, bool rememberPassword) => PostAsync<LoginState>("/api/accounts/login", new { username, password, remember_password = rememberPassword });
    public Task<LoginState> LoginInputAsync(string answer) => PostAsync<LoginState>("/api/accounts/login/input", new { answer });
    public Task<LoginState> ReloginAsync(string username) => PostAsync<LoginState>("/api/accounts/relogin", new { username });
    public Task LogoutAsync(string username) => PostAsync<JsonElement>("/api/accounts/logout", new { username });
    public Task SelectAccountAsync(string username) => PostAsync<JsonElement>("/api/accounts/select", new { username });
    public Task<ParsedTarget> ParseAsync(string url) => PostAsync<ParsedTarget>("/api/parse", new { url });
    public Task<SteamAppInfo> AppInfoAsync(string appId) => GetAsync<SteamAppInfo>($"/api/appinfo/{Uri.EscapeDataString(appId)}");
    public Task<LibraryStatus> LibraryStatusAsync(string username) => GetAsync<LibraryStatus>($"/api/library/status?username={Uri.EscapeDataString(username)}");
    public Task<LibraryStatus> LibrarySyncAsync(string username, bool full) => PostAsync<LibraryStatus>("/api/library/sync", new { username, mode = full ? "full" : "incremental", force_full_sync = full });

    public async Task<List<Job>> JobsAsync()
    {
        var response = await GetAsync<JobsResponse>("/api/jobs");
        return response?.Jobs ?? [];
    }

    public async Task<Job?> JobAsync(string jobId)
    {
        try { return await GetAsync<Job>($"/api/jobs/{Uri.EscapeDataString(jobId)}"); }
        catch (ApiException ex) when (ex.Status == 404) { return null; }
    }

    public async Task<Job> CreateJobAsync(string kind, string id, string username, string os, string depot, string outputDir, string installDir, string name)
    {
        var data = await PostElementAsync("/api/jobs", new
        {
            kind,
            id,
            username,
            anonymous = false,
            os,
            depot,
            output_dir = outputDir,
            install_dir = installDir,
            installdir = installDir,
            name,
        });
        if (data.TryGetProperty("job", out var jobElement))
        {
            return jobElement.Deserialize<Job>(JsonOptions)!;
        }
        return data.Deserialize<Job>(JsonOptions)!;
    }

    public Task JobActionAsync(string jobId, string action) => PostAsync<JsonElement>($"/api/jobs/{Uri.EscapeDataString(jobId)}/{action}", new { });
    public Task JobInputAsync(string jobId, string answer) => PostAsync<JsonElement>($"/api/jobs/{Uri.EscapeDataString(jobId)}/input", new { answer });
    public Task DeleteJobAsync(string jobId, bool deleteFiles) => SendAsync<JsonElement>(HttpMethod.Delete, $"/api/jobs/{Uri.EscapeDataString(jobId)}", new { delete_files = deleteFiles });
    public Task DeleteAllJobsAsync(bool deleteFiles) => SendAsync<JsonElement>(HttpMethod.Delete, "/api/jobs", new { delete_files = deleteFiles });
    public Task<AppSettings> SettingsAsync() => GetAsync<AppSettings>("/api/settings");
    public Task<AppSettings> SaveSettingsAsync(AppSettings settings) => PostAsync<AppSettings>("/api/settings", settings);

    async Task<T> GetAsync<T>(string path, CancellationToken cancellationToken = default) => await SendAsync<T>(HttpMethod.Get, path, null, cancellationToken);
    async Task<T> PostAsync<T>(string path, object body, CancellationToken cancellationToken = default) => await SendAsync<T>(HttpMethod.Post, path, body, cancellationToken);
    async Task<JsonElement> PostElementAsync(string path, object body, CancellationToken cancellationToken = default) => await SendAsync<JsonElement>(HttpMethod.Post, path, body, cancellationToken);

    async Task<T> SendAsync<T>(HttpMethod method, string path, object? body, CancellationToken cancellationToken = default)
    {
        using var request = new HttpRequestMessage(method, path);
        if (body is not null)
        {
            request.Content = new StringContent(JsonSerializer.Serialize(body, JsonOptions), Encoding.UTF8, "application/json");
        }

        using var response = await _client.SendAsync(request, cancellationToken);
        var text = await response.Content.ReadAsStringAsync(cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            throw new ApiException((int)response.StatusCode, ExtractError(text, response.StatusCode));
        }
        if (typeof(T) == typeof(JsonElement) && string.IsNullOrWhiteSpace(text))
        {
            return (T)(object)JsonDocument.Parse("{}").RootElement.Clone();
        }
        return JsonSerializer.Deserialize<T>(string.IsNullOrWhiteSpace(text) ? "{}" : text, JsonOptions)!;
    }

    static string ExtractError(string text, HttpStatusCode statusCode)
    {
        try
        {
            using var doc = JsonDocument.Parse(text);
            if (doc.RootElement.TryGetProperty("error", out var error)) return error.GetString() ?? $"HTTP {(int)statusCode}";
        }
        catch
        {
        }
        return $"HTTP {(int)statusCode}";
    }

    public void Dispose() => _client.Dispose();
}
